### ISSUE TITLE
custom webservice response in picklist

### ISSUE THREAD

**Issue by matizk144** [Microsoft/vso-agent-tasks#667]
Hi!
I have problem about set sourcedefinition in my new task, which call to my webservice. Always I get error when I choose my task in web build creator. Communications was captured using Fiddler.
Example:
I have service, which respond:
{
   "count": 1,
   "value": [   {
      "Id": 1,
      "DisplayName": "mBank_WP",
      "name": "kupsko"
   }]
}

For this I'm trying to set sourcedefinition to parse it:
    "sourceDefinitions": [ 
     { 
         "target": "ProjectsList", 
         "endpoint": "http://192.168.0.175/WindowsMobileDeployer/api/project/tfsformat",
      "selector": "jsonpath:$.value[*].Id",
      "authKey": "tfs:DevTestLabs" 
     } 

First question: What is authkey? For my webservice I don't need any authentication but when I miss authKey request looks like:
{"url":"http://192.168.0.175/WindowsMobileDeployer/api/project/tfsformat","selector":"jsonpath:$.value[*].Id","connectionId":"","scope":"6cf6c3a3-930f-4434-9b69-d48fcb79f401","taskId":"46de93c8-ab19-43e1-a0c9-eebdbd70a5b8"}
and webservice respond
{"$id":"1","innerException":null,"message":"The string must have at least one character.\r\nParameter name: ConnectionId","typeName":"System.ArgumentException, mscorlib, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089","typeKey":"ArgumentException","errorCode":0,"eventId":0}

If authkey is needed, which should I set?, because for exist authkey request looks like:
{"url":"http://192.168.0.175/WindowsMobileDeployer/api/project/tfsformat","selector":"jsonpath:$.value[*].Id","connectionId":"tfs:0000000e-0000-8888-8000-000000000000","scope":"6cf6c3a3-930f-4434-9b69-d48fcb79f401","taskId":"46de93c8-ab19-43e1-a0c9-eebdbd70a5b8"}
but werbservice respond 500 error with json:
{"count":1,"value":{"Message":"An error has occurred."}}

This response looks, that response from my webservice was parsed (should be one item in list), but value was filled kind of errors.

Any ideas?

**Comment by roniga** [comment]
Hi,
   'authKey' refers to the authentication mechanism which would be used to authenticate the request to endpoint (in your case, it would be call to http://192.168.0.175/WindowsMobileDeployer/api/project/tfsformat). 
When the authkey contains "Tfs:xyz" string, it means we want to query "xyz" service in TFS/VSOnline. For external requests, we support only Basic Auth right now. So if your endpoint supports basic auth, you would need to do following to query it:
1. Create a new Generic Connected Service, by going to:
   - Project Settings(Click on the settings button on the top right corner of the screen) -> Services -> Endpoints -> New Service Endpoint -> Generic.
1.1 Enter your Server URL, username, password.
2. Now in your custom task, keep one of the input type as connectedService:Generic. This is where you will be able to select the your connected service, created in step 1.
3. Create your picklist input
4. For your picklist input type, create a source definition, as you had already created. Put the authtype as "$(InputNameForGenericConnectedService)". So for eg., in step 2, if you created the input as:
{
      "name": "ConnectedServiceName",
      "type": "connectedService:Generic",
      "label": "Connected Service",
      "defaultValue": "",
      "required": true,
      "helpMarkDown": "Connected Service"
    },
your sourceDefinition would look something like:
{
        "target": "Step3InputName",
        "endpoint": "Step1ServerUrl/ApiPath",
        "selector": "jsonpath:$.value[*].Id",
        "authKey": "$(ConnectedServiceName)"
    }

Do let me know if there are some more issues.

**Comment by matizk144** [comment]
Thanks for reply, but I still have problem. I created connection in project settings, I added connectedservice:generic input but still this still doesn't work. Of course, in input ConnectedServiceName I have listed my connections, but when I depend sourcedefinition on this I don't see any attempt to call WS, during change connection (I created two generic connections).
It looks like, that this ignored. Also I tried to hardcode connection name (guid also) in task.json, but without react. I respond error: the same as without authkey (my previous post)
My previous conjectures was wrong about call to WS. TFS doesn't connect to my WS (I connected debugger on this)

I changed ip to my local machine, comparing host with my previous post

I paste whole task,json (maybe I miss something):
{
  "id": "46de93c8-ab19-43e1-a0c9-eebdbd70a5b8",
  "name": "WinMobileDeploy",
  "friendlyName": "Publikowania aplikacji Windows Phone",
  "description": "Publikowanie i instalowanie aplikacji na Windows Phone",
  "author": "MKramarz",
  "helpMarkDown": "Replace with markdown to show in help",
  "category": "Deploy",
  "visibility": [
    "Build",
    "Release"
  ],
  "demands": [ ],
  "version": {
    "Major": "0",
    "Minor": "1",
    "Patch": "0"
  },
  "minimumAgentVersion": "1.83.0",
  "instanceNameFormat": "Publikuj na Windows Phone $(musibyc)",
  "inputs": [
    {
      "name": "PlatformType",
      "type": "pickList",
      "label": "Rodzaj platformy",
      "defaultValue": "",
      "required": true,
      "options": {
        "Win8": "Windows 8",
        "Win10": "Windows 10"
      }
    },
    {
     "name": "ConnectedServiceName",
     "type": "connectedService:Generic",
     "label": "Connected Service",
     "defaultValue": "",
     "required": true,
     "helpMarkDown": "Connected Service"
     },
    {
      "name": "ProjectsList",
      "type": "pickList",
      "label": "Projekt",
      "defaultValue": "",
      "required": true
    }
  ],
    "sourceDefinitions": [ 
     { 
         "target": "ProjectsList", 
         "endpoint": "http://192.168.0.14:6010/api/project/tfsformat",
      "selector": "jsonpath:$.value[*].name",
      "authKey": "$(ConnectedServiceName)"
     } 
    ],
  "execution": {
    "PowerShell": {
      "target": "$(currentDirectory)\\win_deploy.ps1",
      "argumentFormat": "",
      "workingDirectory": "$(currentDirectory)"
    }
  }
}

**Comment by roniga** [comment]
The call to "http://192.168.0.14:6010/api/project/tfsformat" would be made from the TFS server(On-Prem/VSOnline). Hence it won't get captured in fiddler. Can you please verify the following:
- The url "http://192.168.0.14:6010/api/project/tfsformat" is accessible from TFS server(in case of VSOnline, this should be a public url)
- Try hitting the above url from an accessible location and see if you are able to access the data with Basic Authentication. Additional help for this : This would basically get translated into a GET request. You can use Fiddler/browser plugins to generate a rest request to your url. For generating basic authentication, this link should help : https://en.wikipedia.org/wiki/Basic_access_authentication#Client_side

**Comment by roniga** [comment]
Also, in case url is accessible, try to use http://192.168.0.14:6010/api/project/tfsformat#$(ConnectedServiceName). This would lead to refreshing of picklist, in case you update your connected service (Empty -> some value -> some other value).

**Comment by matizk144** [comment]
THANKS THANKS THANKS!!!
Finally I resolved the problem. Solution is very easy, that in endpoint it should add #$(ConnectedServiceName), so for example: sourdeDefinition should looks:
    "sourceDefinitions": [ 
     { 
         "target": "ProjectsList", 
         "endpoint": "http://192.168.0.14:6010/api/project/tfsformat#$(ConnectedServiceName)",
      "selector": "jsonpath:$.value[*].DisplayName",
      "authKey": "$(ConnectedServiceName)"
     } 
    ],

About previous questions:
fiddler I used in computer client (Tfs Web send jsons to tfs server),
I test calling to my WS, that I had attached dubugger to my WS, which was hosted on my computer. Tfs call to my computer (my WS).
